File: EulerProjectClassLibrary/EulerProjectClassLibrary/LongSubtraction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Diagnostics;

namespace EulerProjectClassLibrary
{
    /// <summary>
    /// Performs a long subtraction calculation with two numbers.  Does not work when the answer will be negative.
    /// </summary>
    /// <remarks></remarks>
    public class LongSubtraction : EulerClassTwoProperties<string, string, string>
    {
        /// <summary>
        /// Performs a long subtraction calculation with two numbers.  Does not work when the answer will be negative.
        /// </summary>
        /// <param name="a">The larger number.</param>
        /// <param name="b">The smaller number.</param>
        /// <remarks></remarks>
        public LongSubtraction(string a, string b) : base(a, b)
        {
        }

        protected override string Calculate(string a, string b)
        {
            Debug.WriteLine("calculate function started");
            while (a.Length != b.Length) //insert leading 0's onto the start of the shortest number to ensure they fit into the array
            {
                if (a.Length < b.Length) { a = a.Insert(0, "0"); }
                if (b.Length < a.Length) { b = b.Insert(0, "0"); }
                Debug.WriteLine("Add zeros: " + "a: " + a.ToString() + " b: " + b.ToString());
            }

            int[,] array = new int[3, a.Length];

            for (int i = 0; i <= a.Length - 1; i++)
            {
                array[0, i] = Convert.ToInt32(a.Substring(i, 1));
                Debug.WriteLine("Add to array (a): " + a.Substring(i, 1));
                array[1, i] = Convert.ToInt32(b.Substring(i, 1));
                Debug.WriteLine("Add to array (b): " + b.Substring(i, 1));
            }

            for (int i = a.Length - 1; i >= 0; i--)
            {
                int x = array[0, i];
                int y = array[1, i];

                if (x > y) { array[2, i] += (x - y); }
                if (y > x)
                {
                    array[2, i] += 10 - (y - x);
                    if (i != 0)
                    {
                        array[2, i - 1] -= 1;
                    }
                }
            }

            string stranswer = null;
            for (int i = 0; i <= a.Length - 1; i++)
            {
                stranswer += array[2, i].ToString();
                Debug.WriteLine("answer: " + array[2, i].ToString());
            }

            stranswer = stranswer.TrimStart(Convert.ToChar("0"));
            Debug.WriteLine("final answer: " + stranswer);
            return stranswer;
        }

    }
}

File: EulerProjectFunctions/EulerProjectFunctions/LongSubtraction.vb

Public Class LongSubtraction
    Inherits EulerClassTwoProperties(Of String, String, String)

    ''' <summary>
    ''' Performs a long subtraction calculation with two numbers.  Does not work when the answer will be negative.
    ''' </summary>
    ''' <param name="a">The larger number.</param>
    ''' <param name="b">The smaller number.</param>
    ''' <remarks></remarks>
    Sub New(ByVal a As String, ByVal b As String)
        MyBase.New(a, b)
    End Sub

    Protected Overrides Function Calculate(ByVal a As String, ByVal b As String) As Object
        Do Until a.Count = b.Count  'insert leading 0's onto the start of the shortest number to ensure they fit into the array
            If a.Count < b.Count Then a = a.Insert(0, "0")
            If a.Count > b.Count Then b = b.Insert(0, "0")
        Loop

        Dim array(2, a.Count - 1) As Integer 'create array

        For i = 0 To a.Count - 1 'place digits into array
            array(0, i) = a.Substring(i, 1)
            array(1, i) = b.Substring(i, 1)
        Next

        For i = a.Count - 1 To 0 Step -1 'calculate long subtraction, starting from the back
            Dim x As Integer = array(0, i)
            Dim y As Integer = array(1, i)
            If x > y Then
                array(2, i) += x - y
            ElseIf y > x Then
                array(2, i) += 10 - (y - x)
                If i <> 0 Then
                    array(2, i - 1) -= 1
                End If
            End If
        Next

        Dim answer As String = Nothing
        For i = 0 To a.Count - 1
            answer = answer & array(2, i)
        Next
        answer = answer.TrimStart("0")
        If answer = "" Then answer = 0
        Return answer
    End Function
End Class
